### PR TITLE
Add support for resource level locks.

### LINKED ISSUE #3493: [ARM] Lock show command doesn't work
I created a lock on a VM with the following:
`az lock create -g {rg} -n lockvm --resource-type Microsoft.Compute/virtualMachines --resource-name vm1`

The signature for the lock show command is:
```
Command
    az lock show

Arguments
    --name -n [Required]: Name of the lock.
    --resource-group -g : Name of resource group. You can configure the default group using `az
                          configure --defaults group=<name>`.
```

When I tried:

`az lock show -g {rg} -n lockvm`

I get no results. There are no other parameters I could use, even though I know the structure of a lock ID includes the resource that it applies to...

---

### Environment summary

**Install Method:** How did you install the CLI? (e.g. pip, interactive script, apt-get, Docker, MSI, nightly)  
Answer here: dev_setup.py

**CLI Version:** What version of the CLI and modules are installed? (Use `az --version`)  
Answer here: azure-cli (2.0.6+dev)

acr (2.0.4+dev)
acs (2.0.6+dev)
appservice (0.1.6+dev)
batch (3.0.0+dev)
cdn (0.0.2+dev)
cloud (2.0.2+dev)
cognitiveservices (0.1.2+dev)
component (2.0.4+dev)
configure (2.0.6+dev)
core (2.0.6+dev)
cosmosdb (0.1.6+dev)
dla (0.0.6+dev)
dls (0.0.6+dev)
feedback (2.0.2+dev)
find (0.2.2+dev)
interactive (0.3.2+dev)
iot (0.1.5+dev)
keyvault (2.0.4+dev)
lab (0.0.4+dev)
monitor (0.0.4+dev)
network (2.0.6+dev)
nspkg (3.0.0+dev)
profile (2.0.4+dev)
rdbms (0.0.1+dev)
redis (0.2.3+dev)
resource (2.0.6+dev)
role (2.0.4+dev)
sf (1.0.1+dev)
sql (2.0.3+dev)
storage (2.0.6+dev)
taskhelp (0.1.2+dev)
testsdk (0.1.0+dev)
utility-automation (0.1.1)
vm (2.0.6+dev)

Python (Windows) 3.5.1 (v3.5.1:37a07cee5969, Dec  6 2015, 01:38:48) [MSC v.1900 32 bit (Intel)]                                             

**OS Version:** What OS and version are you using?  
Answer here: Win10CU

**Shell Type:** What shell are you using? (e.g. bash, cmd.exe, Bash on Windows)  
Answer here: cmd.exe

### PR DESCRIPTION
**Pull request by brendandburns**
Fixes #3493

### CODE REVIEW

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 1]
```diff
@@ -791,15 +791,41 @@ def _validate_lock_params_match_lock(
...
+    :param parent_resource_path: Path to a parent resource
+    :type parent_resource_path: str
+    :param resource_type: The type for the resource with the lock.
+    :type resource_type: str
+    :param resource_name: Name of a resource that has a lock.
+    :type resource_name: str
     """
     lock_client = _resource_lock_client_factory()
+
+    lock_resource = _extract_lock_params(resource_group_name, resource_provider_namespace,
```
Comment (tjprescott) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: I believe `_extract_lock_params` may be the source of the bug that doesn't allow `--resource-type` to accept <NAMESPACE/TYPE>

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 2]
```diff
@@ -791,15 +791,41 @@ def _validate_lock_params_match_lock(
...
-def get_lock(name, resource_group_name=None):
+def get_lock(name, resource_group_name=None, resource_provider_namespace=None,
+             parent_resource_path=None, resource_type=None, resource_name=None):
     """
-    :param name: Name of the lock.
+    :param name: The name of the lock.
     :type name: str
+    :param resource_provider_namespace: Name of a resource provider.
+    :type resource_provider_namespace: str
+    :param parent_resource_path: Path to a parent resource
```
Comment (tjprescott) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: Help text should either be specified in `_help.py` (probably best) or the `params.py` file. This is the only way of adding help that we actively discourage (it is really meant for SDK reflection). 😇

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 3]
```diff
@@ -791,15 +791,41 @@ def _validate_lock_params_match_lock(
                 name, _resource_name))
 
 
-def get_lock(name, resource_group_name=None):
+def get_lock(name, resource_group_name=None, resource_provider_namespace=None,
```
Comment (tjprescott) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: This would also be the perfect time to fix #3476, as this is a significant change to this command signature.

**Comment by tjprescott** [comment]
The --resource-type parsing has been fixed by #3584 I believe.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py (hunk 1 above)
Comment (brendandburns) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: Fixed by #3584...

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py (hunk 3 above)
Comment (brendandburns) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: done.

**Comment by brendandburns** [comment]
Comments addressed, please re-check.

Thanks!

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_validators.py [hunk 4]
```diff
@@ -38,7 +38,7 @@ def internal_validate_lock_parameters(resource_group_name, resource_provider_nam
             raise CLIError('--resource-type is ignored if --resource-group is not given.')
         if resource_provider_namespace is not None:
             raise CLIError('--namespace is ignored if --resource-group is not given.')
-        if parent_resource_path is not None:
+        if parent_resource_path and parent_resource_path is not None:
```
Comment (tjprescott) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_validators.py]: If `parent_resource_path == None` then `if parent_resource_path` would return false anyways. So you could just simplify this to `if parent_resource_path is not None`.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 5]
```diff
@@ -777,7 +777,7 @@ def _validate_lock_params_match_lock(
             _resource_type = resource.get('child_type', None)
             _resource_name = resource.get('child_name', None)
             parent = (resource['type'] + '/' + resource['name'])
-            if parent != parent_resource_path:
+            if parent and parent != parent_resource_path:
```
Comment (tjprescott) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: This was fine the way it was before.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_validators.py (hunk 4 above)
Comment (brendandburns) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_validators.py]: done.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py (hunk 5 above)
Comment (brendandburns) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: done.

**Comment by tjprescott** [comment]
CI is failing for this test.

**Comment by brendandburns** [comment]
@tjprescott I see:

```
CLIError: Please run 'az login' to setup account.
```

in my scenario test. Why is that there? Do I really need to run `az login` in the test?

**Comment by tjprescott** [comment]
No, you don't.  I would try re-recording the test and see if that fixes it. You'll get that error if there's no recording because the CI will try to run the test live and fail. You could also get it if you had a bad recording which recorded that as the response.

**Comment by tjprescott** [comment]
@brendandburns can you get the CI passing so this can be merged?  I'm looking to add support for `--ids` to these commands and would like to get this PR in beforehand.

**Comment by brendandburns** [comment]
@tjprescott recorded tests failed again. These things are annoying to create and super brittle, can you see if you can tell what the problem is?

Thanks

**Comment by tjprescott** [comment]
@brendandburns I would suggest creating the lock on a VNet instead of a storage account--it will make it easier to record since VNet name doesn't have to be globally unique.  However, I don't think that is the source of the issue.

There may be an issue with ScenarioTest that is causing this.  You could try writing the test with the older test base (ResourceGroupVCRTestBase) and see if that works. ScenarioTest is newer and there are edge cases the old test base covered that the newer version doesn't. 

@troydai do you have any ideas on this one?

**Comment by troydai** [comment]
We replace the subscription id in the recording to erase credentials. And the test failed because the full zero subscription id gets in the way. I don't know what exactly happen util I can debug into the code.

**Comment by brendandburns** [comment]
@tjprescott you requested the scenario test for this PR, if you'd like to merge this, I can move the scenario test out to a different PR...

**Comment by tjprescott** [comment]
@brendandburns that sounds perfect.  Please just comment the test out, then we can merge and we can re-enable and debug into it.

**Comment by brendandburns** [comment]
@tjprescott I think this is ready...

**Comment by brendandburns** [comment]
(ah, nm, sorry github hadn't updated to see the merge...)
